contracts/TWFee.sol

```diff
@@ -53,99 +85,126 @@ contract TWFee is Multicall, ERC2771Context, AccessControlEnumerable {
+        
+        FeeInfo memory targetFeeInfo = feeInfo[tierToUse][_feeType];
+        (recipient, bps) = (targetFeeInfo.recipient, targetFeeInfo.bps);
+    }
+
+    /// @dev Lets an approved caller select a subscription for `_for`.
+    function selectSubscription(
+        address _for,
+        uint256 _tier,
+        uint256 _priceToPay,
+        address _currencyToUse
+    )
+        external 
+        payable
+    {
+        address caller = _msgSender();
+        require(
+            _for == caller || isApproved[_for][caller] || hasRole(DEFAULT_ADMIN_ROLE, caller),
+            "not approved to select tier."
+        );
+
+        uint256 durationForTier = tierInfo[_tier].duration;
+        require(durationForTier != 0, "invalid tier.");
+
+        bool isValidPaymentInfo = tierInfo[_tier].isCurrencyApproved[_currencyToUse] || tierInfo[_tier].priceForCurrency[_currencyToUse] == _priceToPay;
+        require(isValidPaymentInfo, "invalid payment info.");
+
+        Tier memory tierSelected = Tier({
+            tier: uint128(_tier),
+            validUntilTimestamp: uint128(block.timestamp + durationForTier)
         });
 
-        _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
-        _setupRole(FEE_ROLE, _msgSender());
-    }
+        tierForUser[_for] = tierSelected;
+
+        CurrencyTransferLib.transferCurrency(_currencyToUse, _msgSender(), thirdwebTreasury, _priceToPay);
 
-    function getFeeInfo(address _module, FeeType _feeType) external view returns (address recipient, uint256 bps) {
-        bytes32 moduleType = IThirdwebModule(_module).moduleType();
-
-        FeeInfo memory infoForModuleInstance = feeInfoByModuleInstance[_module][_feeType];
-        FeeInfo memory infoForModuleType = feeInfoByModuleType[moduleType][_feeType];
-        FeeInfo memory defaults = defaultFeeInfo[_feeType];
-
-        // Get appropriate fee bps
-        bps = infoForModuleInstance.bps;
-        if (bps == 0) {
-            bps = infoForModuleType.bps;
-        }
-        if (bps == 0) {
-            bps = defaults.bps;
-        }
-
-        // Get appropriate fee recipient
-        recipient = infoForModuleInstance.recipient;
-        if (recipient == address(0)) {
-            recipient = infoForModuleType.recipient;
-        }
-        if (recipient == address(0)) {
-            recipient = defaults.recipient;
-        }
+        emit TierForUser(_for, _tier, _currencyToUse, _priceToPay, tierSelected.validUntilTimestamp);
     }
 
-    /// @dev Lets the admin set fee bps and recipient for the given module type and fee type.
-    function setFeeInfoForModuleType(
-        bytes32 _moduleType,
-        uint256 _feeBps,
-        address _feeRecipient,
-        FeeType _feeType
-    ) external onlyFeeAdmin onlyValidFee(_feeBps) {
-        FeeInfo memory feeInfo = FeeInfo({ bps: _feeBps, recipient: _feeRecipient });
+    /// @dev Lets the caller renew a subscription for `_for`.
+    function renewSubscription(
```

Comment:
I'm not sure if it's a good idea to mix in subscription stuffs into TWFee.
I think we should make TWFee agnostic to how the payment / subscription is being done.
Have a separate contract that take care of just subscriptions. With payments, we'd probably want clean & clear separation for accounting raeson.